category_theory/category.lean

```diff
@@ -0,0 +1,44 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Stephen Morgan, Scott Morrison
+
+import tactic.make_lemma
+import tactic.interactive
+
+namespace categories
+
+universes u v
+
+meta def obviously := `[skip]  
+
+/- 
+The propositional fields of `category` are annotated with the auto_param `obviously`, which is just a synonym for `skip`.
+Actually, there is a tactic called `obviously` which is not part of this pull request, which should be used here. It successfully
+discharges a great many of these goals. For now, proofs which could be provided entirely by `obviously` (and hence omitted entirely
+and discharged by an auto_param), are all marked with a comment "-- obviously says:".
+-/
+
+class category (Obj : Type u) : Type (max u (v+1)) :=
+  (Hom : Obj → Obj → Type v)
+  (identity : Π X : Obj, Hom X X)
+  (compose  : Π {X Y Z : Obj}, Hom X Y → Hom Y Z → Hom X Z)
+  (left_identity  : ∀ {X Y : Obj} (f : Hom X Y), compose (identity X) f = f . obviously)
+  (right_identity : ∀ {X Y : Obj} (f : Hom X Y), compose f (identity Y) = f . obviously)
+  (associativity  : ∀ {W X Y Z : Obj} (f : Hom W X) (g : Hom X Y) (h : Hom Y Z), compose (compose f g) h = compose f (compose g h) . obviously)
+
+notation `𝟙` := category.identity     -- type as \b1
```

Comment:
I can imagine that we will also love to have this notation for the unit object in a monoidal category. But then... we also want good notation for identity morphisms, and I don't have a good alternative in mind.

Reply:
Besides the double-struck 1, there are several other variant numeral 1s available in unicode:

"bold": 𝟏	
"sans-serif": 𝟣
"sans-serif bold": 𝟭
"monospace": 𝟷

Currently none of these have shortcuts in VSCode, but those are easy to add. My monoidal categories library hasn't fixed on any notation for monoidal units. I'm not sure that having two slightly different 1s is a good idea.